### PR TITLE
Add Services Support for EKS and EMR

### PR DESCRIPTION
**Pull request by bwagner5**
Issue #, if available:
https://github.com/aws/amazon-ec2-instance-selector/issues/68

Description of changes:

#### TL;DR:
This PR adds support for arbitrary service filtering based on a simple service interface which transforms the user supplied filters (just like the aggregate filter functionality).  This PR adds two implementations of services for EMR and EKS w/ versioning. 

Example CLI usage:

```
$ ec2-instance-selector --service eks --max-results 5
a1.2xlarge
a1.4xlarge
a1.large
a1.medium
a1.metal
NOTE: 374 entries were truncated, increase --max-results to see more

$ ec2-instance-selector --service eks-20210125 --max-results 5
a1.2xlarge
a1.4xlarge
a1.large
a1.medium
a1.metal
NOTE: 374 entries were truncated, increase --max-results to see more

$ ec2-instance-selector --service emr --max-results 5
c1.medium
c1.xlarge
c3.2xlarge
c3.4xlarge
c3.8xlarge
NOTE: 191 entries were truncated, increase --max-results to see more

$ ec2-instance-selector --service emr-5.25.0 --max-results 5
c1.medium
c1.xlarge
c3.2xlarge
c3.4xlarge
c3.8xlarge
NOTE: 197 entries were truncated, increase --max-results to see more
```

#### Details:

 - InstanceTypes was added as a Filter so that service filters can start with a universal set of instance types without restricting a user's use of the allow-list. 
 -  VirtType was added as a filter since emr and eks both only support hvm instance types

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/selector/eks.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+const (
+	eksAMIRepoURL               = "https://github.com/awslabs/amazon-eks-ami"
+	eksFallbackLatestAMIVersion = "v20210125"
```
Comment (brycahta) [inline pkg/selector/eks.go:17]: np: would be neat to fetch EKS AMI based on k8s major/minor versions per this [table](https://docs.aws.amazon.com/eks/latest/userguide/eks-linux-ami-versions.html)

File: pkg/selector/emr.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	filters.RootDeviceType = aws.String("ebs")
+	filters.VirtualizationType = aws.String("hvm")
+	return filters, nil
+}
+
+// getEMRInstanceTypes returns a list of instance types that emr supports
+func (e EMR) getEMRInstanceTypes(version string) ([]string, error) {
+	instanceTypes := []string{}
+	if version == "" {
+		version = "5.20.0"
```
Comment (brycahta) [inline pkg/selector/emr.go:30]: np: `const`

also is this latest, fallback, or both?

File: pkg/selector/emr.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+	for _, instanceType := range e.getAllEMRInstanceTypes() {
+		if versionInt >= 5250 {
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 5200 && versionInt < 5250 {
+			if e.isOnlyEMR_5_25_0_plus(instanceType) {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		} else if versionInt >= 5150 && versionInt < 5200 {
+			if instanceType == "c1.medium" {
```
Comment (brycahta) [inline pkg/selector/emr.go:46]: why call out this type?

File: pkg/selector/emr.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+			if strings.HasPrefix(instanceType, "i3") {
+				continue
+			}
+			instanceTypes = append(instanceTypes, instanceType)
+		}
+	}
+	return instanceTypes, nil
+}
+
+func (EMR) isEMR_5_13_0_plus(instanceType string) bool {
```
Comment (brycahta) [inline pkg/selector/emr.go:94]: nice naming 👍

File: pkg/selector/emr.go [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+	for _, prefix := range prefixes {
+		if strings.HasPrefix(instanceType, prefix) {
+			return true
+		}
+	}
+	return false
+}
+
+func (EMR) getAllEMRInstanceTypes() []string {
+	return []string{
```
Comment (brycahta) [inline pkg/selector/emr.go:138]: how often will this need to be updated?

File: pkg/selector/services_test.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	registry.RegisterAWSServices()
+
+	emr := ""
+	filters := selector.Filters{
+		Service: &emr,
+	}
+
+	transformedFilters, err := registry.ExecuteTransforms(filters)
+	h.Ok(t, err)
+	h.Assert(t, transformedFilters == filters, " Filters should no be modified")
```
Comment (brycahta) [inline pkg/selector/services_test.go:97]: np: not*

File: cmd/main.go [hunk 7]
```diff
@@ -121,29 +123,31 @@ Full docs can be found at github.com/aws/amazon-` + binName
...
 	cli.IntMinMaxRangeFlags(networkPerformance, nil, nil, "Bandwidth in Gib/s of network performance (Example: 100)")
 	cli.RegexFlag(allowList, nil, nil, "List of allowed instance types to select from w/ regex syntax (Example: m[3-5]\\.*)")
 	cli.RegexFlag(denyList, nil, nil, "List of instance types which should be excluded w/ regex syntax (Example: m[1-2]\\.*)")
+	cli.StringOptionsFlag(virtualizationType, nil, nil, "Virtualization Type supported: [hvm or pv]", []string{"hvm", "paravirtual", "pv"})
 
 	// Suite Flags - higher level aggregate filters that return opinionated result
 
 	cli.SuiteStringFlag(instanceTypeBase, nil, nil, "Instance Type used to retrieve similarly spec'd instance types", nil)
 	cli.SuiteBoolFlag(flexible, nil, nil, "Retrieves a group of instance types spanning multiple generations based on opinionated defaults and user overridden resource filters")
+	cli.SuiteStringFlag(service, nil, nil, "Filter instance types based on service support (Example: eks, eks-20201211, or emr-5.20.0)", nil)
```
Comment (brycahta) [inline cmd/main.go:150]: out of curiosity-- is the versioning for eks/emr well known? Or is it more along the lines of, if someone is using this tool for one of these services, then they would know?

File: pkg/selector/eks.go (hunk 1 above)
Comment (bwagner5) [reply pkg/selector/eks.go:17]: I think we'd still need the date though and the instance types supported don't change across major.minor k8s versions. So doing that would only require users to input more information not less, but we probably should support the full version as an option of input. We can just pull off the date at the end to actually do the lookup.

File: pkg/selector/emr.go (hunk 2 above)
Comment (bwagner5) [reply pkg/selector/emr.go:30]: both, we'll fallback to the latest.  👍  on const, I'll make that change

File: pkg/selector/emr.go (hunk 3 above)
Comment (bwagner5) [reply pkg/selector/emr.go:46]: I just called it that since there was already a string version and this should be the string version w/ no dot as an int for comparisons. Definitely not the cleanest.... maybe a better way is to make a comparison func that accepts the version as a string.

File: pkg/selector/emr.go (hunk 4 above)
Comment (bwagner5) [reply pkg/selector/emr.go:94]: .... 😹  .... 🙈

File: pkg/selector/emr.go (hunk 5 above)
Comment (bwagner5) [reply pkg/selector/emr.go:138]: I'm not sure.... This is definitely not an optimal solution, but I think it'll be the best hack available. Other downsides are that users will have to actually update the ec2-instance-selector binary to get updated versions or to add supported instance types to older versions.

File: pkg/selector/services_test.go (hunk 6 above)
Comment (bwagner5) [reply pkg/selector/services_test.go:97]: 👍

File: cmd/main.go (hunk 7 above)
Comment (bwagner5) [reply cmd/main.go:150]: It's probably the latter, if someone is using this tool to provision those service's resources, then they'll probably know.  But if not, we'll still support just the service name without version to default to the latest or some default that makes sense for the service.

File: pkg/selector/emr.go (hunk 2 above)
Comment (bwagner5) [reply pkg/selector/emr.go:30]: I realized this is not the latest... I think this is the most utilized, I'll make const reflect that it's a fallback and not latest.

File: pkg/selector/emr.go (hunk 3 above)
Comment (bwagner5) [reply pkg/selector/emr.go:67]: used semver lib instead

File: pkg/selector/emr.go [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+// EMR is a Service type for a custom service filter transform
+type EMR struct{}
+
+// Filters implements the Service interface contract for EMR
+func (e EMR) Filters(version string) (Filters, error) {
+	filters := Filters{}
+	if version == "" {
+		version = fallbackVersion
+	}
+	semanticVersion, err := semver.Make(version)
```
Comment (brycahta) [inline pkg/selector/emr.go:37]: oo nice 🤩

**Review by brycahta — APPROVED** [review]
Looks great! 👍

File: pkg/selector/emr.go (hunk 5 above)
Comment (brycahta) [reply pkg/selector/emr.go:159]: Gotcha, if we see ourselves needing to manually update this list often then we could look into watching and/or pulling directly from the [github readme on emr types](https://github.com/awsdocs/amazon-emr-management-guide/blob/master/doc_source/emr-supported-instance-types.md)

**Comment by rothgar** [comment]
Looks great! I was able to combine `--service eks` with `-o one-line` for what I needed with `eksctl` :+1:
